src/SmartEnum/Core/SmartEnumWhen.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Ardalis.SmartEnum.Core
+{
+    public readonly struct SmartEnumWhen<TEnum, TValue>
+        where TEnum : SmartEnum<TEnum, TValue>
+        where TValue : IEquatable<TValue>, IComparable<TValue>
+    {
+        private readonly SmartEnum<TEnum, TValue> smartEnum;
+        private readonly bool stopEvaluating;
+
+        public SmartEnumWhen(bool stopEvaluating, SmartEnum<TEnum, TValue> smartEnum)
```

Comment:
The constructor can optionally be set to `internal` so that instances can only be created by the methods `When` and `Then`.